x/auth/pubkey/handler.go

```diff
@@ -0,0 +1,56 @@
+package pubkey
+
+import (
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+	"github.com/cosmos/cosmos-sdk/x/auth/keeper"
+	"github.com/cosmos/cosmos-sdk/x/auth/pubkey/types"
+	authtypes "github.com/cosmos/cosmos-sdk/x/auth/types"
+)
+
+// NewHandler returns a handler for x/auth message types.
+func NewHandler(ak keeper.AccountKeeper, bk types.BankKeeper, sk types.StakingKeeper) sdk.Handler {
+	return func(ctx sdk.Context, msg sdk.Msg) (*sdk.Result, error) {
+		ctx = ctx.WithEventManager(sdk.NewEventManager())
+
+		switch msg := msg.(type) {
+		case *types.MsgChangePubKey:
+			return handleMsgChangePubKey(ctx, ak, bk, sk, msg)
+
+		default:
+			return nil, sdkerrors.Wrapf(sdkerrors.ErrUnknownRequest, "unrecognized %s message type: %T", types.ModuleName, msg)
+		}
+	}
+}
+
+func handleMsgChangePubKey(ctx sdk.Context, ak keeper.AccountKeeper, bk types.BankKeeper, sk types.StakingKeeper, msg *types.MsgChangePubKey) (*sdk.Result, error) {
+	acc := ak.GetAccount(ctx, msg.Address)
+	if acc == nil {
+		return nil, sdkerrors.Wrapf(sdkerrors.ErrInvalidRequest, "account %s does not exist", msg.Address)
+	}
+	acc.SetPubKey(msg.PubKey)
+	ak.SetAccount(ctx, acc)
+
+	// handle additional fee logic inside MsgChangePubKey handler
+	signers := msg.GetSigners()
+	if len(signers) == 0 {
+		return nil, sdkerrors.Wrapf(sdkerrors.ErrInvalidRequest, "signers should exist")
+	}
+	feePayer := signers[0]
+	amount := ak.GetParams(ctx).PubKeyChangeCost // should get from auth params
+	fees := sdk.Coins{sdk.NewInt64Coin(sk.BondDenom(ctx), int64(amount))}
```

Comment:
You don't need to charge additional fees like this.  Sorry, what I meant was to just charge some additional gas (which you do correctly further down on line 46).  That way additional fees will just get automatically deducted, by proxy of a higher gas cost.

Reply:
Although, what is nice about this way is that fees are mandatorily deducted, while currently in the Cosmos SDK, there's no in-consensus minimum gas price, and so often the gas price is just 0 right now because a few validators charge 0.

(But this is a problem that should be fixed, maybe that's the next thing we tackle 😄)